gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,382 @@
+
+		private int level;
+
+		// Represents the boundary of the node
+		private final BoundingBox bounds;
+		private final Array<OctreeNode> children = new Array<>(8);
+		private final Array<T> geometries = new Array<T>(maxItemsPerNode);
+
+		public OctreeNode () {
+			bounds = new BoundingBox();
+		}
+
+		private void split () {
+			float midx = (bounds.max.x + bounds.min.x) * 0.5f;
+			float midy = (bounds.max.y + bounds.min.y) * 0.5f;
+			float midz = (bounds.max.z + bounds.min.z) * 0.5f;
+
+			int deeperLevel = level - 1;
+
+			children.add(createNode(new Vector3(bounds.min.x, midy, midz), new Vector3(midx, bounds.max.y, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, midy, midz), new Vector3(bounds.max.x, bounds.max.y, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, midy, bounds.min.z), new Vector3(bounds.max.x, bounds.max.y, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, midy, bounds.min.z), new Vector3(midx, bounds.max.y, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, bounds.min.y, midz), new Vector3(midx, midy, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, bounds.min.y, midz), new Vector3(bounds.max.x, midy, bounds.max.z), deeperLevel));
+			children.add(createNode(new Vector3(midx, bounds.min.y, bounds.min.z), new Vector3(bounds.max.x, midy, midz), deeperLevel));
+			children.add(createNode(new Vector3(bounds.min.x, bounds.min.y, bounds.min.z), new Vector3(midx, midy, midz), deeperLevel));
+
+			// Move geometries from parent to children
+			for (OctreeNode child : children) {
+				for (T geometry : this.geometries) {
+					child.add(geometry);
+				}
+			}
+			this.geometries.clear();
+		}
+
+		private void merge() {
+			for (OctreeNode node : children) {
+				geometries.addAll(node.geometries);
+				freeNode(node);
+			}
+			children.clear();
+		}
+
+		private OctreeNode createNode(Vector3 min, Vector3 max, int level) {
+			OctreeNode node = nodePool.obtain();
+			node.bounds.set(min, max);
+			node.level = level;
+			return node;
+		}   
+
+		private void freeNode(OctreeNode node) {
+			node.geometries.clear();
+			node.children.clear();
+			nodePool.free(node);
+		}
+
+		protected boolean contains (Collider<T> boundingBoxHandler, T geometry) {
+			return boundingBoxHandler.intersects(bounds, geometry);
+		}
+
+		protected void add (T geometry) {
+			if (!collider.intersects(bounds, geometry)) {
+				return;
+			}
+
+			// If is not leaf, check children
+			if (!isLeaf()) {
+				for (OctreeNode child : children) {
+					child.add(geometry);
+				}
+			} else {
+				geometries.add(geometry);
+				if (geometries.size > maxItemsPerNode && level > 0) {
+					split();
+				}
+			}
+		}
+
+		protected boolean remove (T object) {
+			if (!isLeaf()) {
+				boolean removed = false;
+				int childrenSum = 0;
+				for (OctreeNode node : children) {
+					childrenSum += node.geometries.size;
```

Comment:
size may change after removing, i think you can swap those lines.

Reply:
You are right! Done.